File: Baekjun/8.4/3033. 가장 긴 문자열/solution.swift
func LCP(_ sa:[Int], _ rank:[Int], _ n:Int, _ S:String) -> [Int] {
  var lcp:[Int] = Array(repeating: 0, count: n)
  var k = 0
  let S = Array(S)

  for i in 0..<n {
    if rank[i] == 0 {
      continue
    }
    
    let j = sa[rank[i] - 1]

    if k > 0 {
      k -= 1
    }

    while j + k < n && i + k < n && S[j + k] == S[i + k] {
      k += 1
    }

    lcp[rank[i]] = k
  }
  
  return lcp
}

func MM(_ n:Int, _ S:String) -> [Int] {
  var sa:[Int] = Array(0..<n)
  var rank:[Int] = Array(S).compactMap { Int($0.asciiValue!) }
  var tmp:[Int] = Array(repeating:0, count: n)
  var k = 1

  while k <= n {
    let cmp:[Int] = (0..<n).map {
      if $0 + k < n {
        return rank[$0] << 20 | (1 + rank[$0 + k] )
      }
      return rank[$0] << 20 | 0
    }

    sa.sort {
      cmp[$0] < cmp[$1]
    }
    
    tmp[sa[0]] = 0

    for i in 1..<n {
      tmp[sa[i]] = tmp[sa[i-1]] + (cmp[sa[i]] > cmp[sa[i-1]] ? 1 : 0)
    }
    
    rank = tmp.map{$0}
    k <<= 1
    
  }
  
  return LCP(sa, rank, n, S)
}

let maxVal = MM(Int(readLine()!)!, readLine()!).max()!

if maxVal >= 2 {
  print(maxVal)
} else {
  print(0)
}

File: Baekjun/8.4/3033. 가장 긴 문자열/solution.py
def lcp(sa, rank, n, S):
    lcp = [0] * n
    k = 0

    for i in range(n):
        j = sa[rank[i] - 1]

        if k:
            k -= 1

        while j + k < n and i + k < n and S[j + k] == S[i + k]:
            k += 1

        lcp[rank[i]] = k

    return lcp


def MM(n, S):
    sa = list(range(n))
    rank = list(map(ord, S))
    tmp = [0] * n
    k = 1

    while k <= n:
        cmp = [rank[i] << 20 | (1 + rank[i + k] if i + k < n else 0) for i in range(n)]

        sa.sort(key=lambda i: cmp[i])

        tmp[sa[0]] = 0

        for i in range(1, n):
            tmp[sa[i]] = tmp[sa[i - 1]] + (cmp[sa[i - 1]] < cmp[sa[i]])

        rank = tmp[:]
        k <<= 1

    return lcp(sa, rank, n, S)


max_val = max(MM(int(input()), input()))

if max_val >= 2:
    print(max_val)
else:
    print(0)
